apps/evm/lib/evm/gas.ex

```diff
@@ -134,36 +134,50 @@ defmodule EVM.Gas do
 
   ## Examples
 
-      # TODO: Figure out how to hand in state
-      iex> EVM.Gas.cost(%EVM.MachineState{}, %EVM.ExecEnv{})
+      iex> {_exec_env, cost} = EVM.Gas.cost(%EVM.MachineState{}, %EVM.ExecEnv{})
+      iex> cost
       0
   """
-  @spec cost(MachineState.t(), ExecEnv.t()) :: t
+  @spec cost(MachineState.t(), ExecEnv.t()) :: {ExecEnv.t(), t()}
   def cost(machine_state, exec_env) do
-    case cost_with_status(machine_state, exec_env) do
-      {:original, cost} -> cost
-      {:changed, value, _} -> value
-    end
+    {updated_exec_env, cost_with_status} = cost_with_status(machine_state, exec_env)
+
+    cost =
+      case cost_with_status do
+        {:original, cost} -> cost
+        {:changed, value, _} -> value
+      end
+
+    {updated_exec_env, cost}
   end
 
-  @spec cost_with_status(MachineState.t(), ExecEnv.t()) :: cost_with_status()
+  @spec cost_with_status(MachineState.t(), ExecEnv.t()) :: {ExecEnv.t(), cost_with_status()}
   def cost_with_status(machine_state, exec_env) do
     operation = MachineCode.current_operation(machine_state, exec_env)
     inputs = Operation.inputs(operation, machine_state)
-    operation_cost = operation_cost(operation.sym, inputs, machine_state, exec_env)
+
+    {updated_exec_env, operation_cost} =
+      case operation_cost(operation.sym, inputs, machine_state, exec_env) do
```

Comment:
Since we're passing in `exec_env` to `operation_cost`, it's surprising that it either returns just the cost or a tuple of the cost and exec env. Would it be possible to remove this case statement by having `operation_cost` consistently return `{exec_env, cost}`?

Reply:
@MattMSumner most operation cost functions don't modify exec_env. But we can return unmodified exec_env.